lib/ide/src/controller.rs

```diff
@@ -0,0 +1,426 @@
+//! manage their lifetime.
+//!
+//! Primarily views are considered owners of their respective controllers.
+//! Additionally, controllers are allowed to keep strong handle "upwards".
+//!
+//! Controllers store their handles using `utils::cell` handle types to ensure
+//! that mutable state is safely accessed.
+
+use crate::prelude::*;
+
+/// General-purpose `Result` supporting any `Error`-compatible failures.
+pub type FallibleResult<T> = Result<T,failure::Error>;
+
+/// Macro defines `StrongHandle` and `WeakHandle` newtypes for handles storing
+/// the type given in the argument.
+///
+/// This allows treating handles as separate types and fitting them with impl
+/// methods of their own. While not necessary, such implementation may allow
+/// hiding from user gritty details of `with` usage behind nice, easy API.
+macro_rules! make_handles {
+    ($data_type:ty) => {
+        /// newtype wrapper over StrongHandle.
+        #[derive(Shrinkwrap)]
+        #[derive(Clone,Debug)]
+        pub struct StrongHandle(pub utils::cell::StrongHandle<$data_type>);
+
+        impl StrongHandle {
+            /// Obtain a WeakHandle to this data.
+            pub fn downgrade(&self) -> WeakHandle {
+                WeakHandle(self.0.downgrade())
+            }
+            /// Create a new StrongHandle that will wrap given data.
+            pub fn new(data:$data_type) -> StrongHandle {
+                StrongHandle(utils::cell::StrongHandle::new(data))
+            }
+        }
+
+        /// newtype wrapper over WeakHandle.
+        #[derive(Shrinkwrap)]
+        #[derive(Clone,Debug)]
+        pub struct WeakHandle(pub utils::cell::WeakHandle<$data_type>);
+
+        impl WeakHandle {
+            /// Obtain a StrongHandle to this data.
+            pub fn upgrade(&self) -> Option<StrongHandle> {
+                self.0.upgrade().map(StrongHandle)
+            }
+        }
+    };
+}
+
+
+
+// =======================
+// === Text controller ===
+// =======================
+
+pub mod project {
+    //! Project controller.
+    ///
+    /// Responsible for owning any remote connection clients. Expected to live
+    /// as long as the project remains open in the IDE.
+    use super::*;
+    use json_rpc::Transport;
+
+    /// Create a new project controller.
+    ///
+    /// The remote connections should be already established.
+    pub fn new(file_manager_transport:impl Transport + 'static) -> StrongHandle {
+        let file_manager = file_manager_client::Client::new(file_manager_transport);
+        let ret = Data {
+            file_manager,
+            module_cache: default(),
+        };
+        StrongHandle::new(ret)
+    }
+
+    /// Project controller's state.
+    #[derive(Debug)]
+    pub struct Data {
+        /// File Manager Client.
+        file_manager : file_manager_client::Client,
+        /// Cache of module controllers. As the weak is handle (we don't manage
+        /// their lifetime), dead ones might still be included.
```

Comment:
Again, what is a "dead module". After reading the whole file I'd recommend to explain all the important terms in one place, preferably in the module doc.